OpenTrons/opentrons-api
# fix(update-server): Keep name deconflicted with other devices on the network

# Overview

Fixes #10126.

# Changelog

* Expand our code for communicating with the Avahi daemon over D-Bus, to give us the ability to detect when the Avahi service name has collided with another device on the network.
* While `update-server` is running, constantly monitor Avahi for name collisions. When one happens, automatically set a new name, like `My Robot #2`. This is the addition that fixes #10126.
* At all times, keep the machine's 3 human-readable names in sync with each other:
  * The Avahi service name, which Avahi advertises over mDNS + DNS-SD.
  * The name that `update-server` exposes over HTTP via `GET /server/name`, `POST /server/name`, and `GET /server/update`.
  * The *pretty* hostname, which is read [outside of `update-server`](https://github.com/Opentrons/opentrons/blob/8fc1c0c95b38f10eeb3daa1c10bfd07d29782e22/api/src/opentrons/config/__init__.py#L102-L106) and [winds up in `robot-server`'s `/health`endpoint](https://github.com/Opentrons/opentrons/blob/40bc918b82b4de561f8d092708a84966196d21bd/robot-server/robot_server/health/router.py#L39).
  
  All 3 of these names are  used by the discovery client. Keeping them in sync with each other is important to avoid confusing the discovery client, in its current implementation. See #10199.

* The *static*  hostname is unaffected by this PR. It continues to follow the robot's serial number.

# Reviewing

## Recommended code review order

Here is one way to ease into this PR. Working from the outside in:

1. In `name_management/__init__.py`...
    1. Check out the docstring to familiarize yourself with some important terminology.
    2. Check out how the HTTP endpoint functions `set_name_endpoint()` and `get_name_endpoint()` changed.
2. In `name_managemement/name_synchronizer.py`, check out the `NameSynchronizer` interface, which the endpoints now use, and its `RealNameSynchronizer` implementation.
3. In `name_management/avahi.py`, check out `AvahiClient`, a dependency of `NameSynchronizer`.
4. Check out the server initialization code in `buildroot/__main__.py` and `buildroot/__init__.py`. Note that this is heavily duplicated with `openembedded/__main__.py` and `openembedded/__init__.py`.
5. Everything else.

## Testing

### What to test

We need to verify the following properties:

* The robot's DNS-SD instance name is always the same as the name accessible over HTTP.
    * Caveat: The name in `GET /health` may lag behind the one in `GET /server/update/health` sometimes, because of a different bug. See https://github.com/Opentrons/opentrons/issues/10413.
* The name automatically changes to deconflict with other devices on the network

### How to test it

This needs at least one real OT-2.

On macOS, you can synthesize fake devices on the network to create collisions by running:

```bash
dns-sd -R 'My Fake Robot Name' _http._tcp . 31950
```

From there, there are many ways to inspect the OT2s' names:

* You can simply play around on the Opentrons App. You might want to dig into the app's network settings to clear the robot discovery cache to make things cleaner and make it easier to notice anomalies.
* You can use the macOS `dns-sd -B` command to monitor just the DNS-SD instance names. (Pay attention to the `add`/`remove` column!)
* You can use Postman or `curl` to directly check `GET /health`, `GET /server/health`, and `GET /server/name`.
* You can use the Opentrons discovery client command-line interface (`cd discovery-client && yarn run discovery`).
* The macOS `dns-sd -B` command. 

Interesting scenarios to try out:

* Two machines are both online and have different names. You change one of their names so that they match. One of them should notice the conflict and correct it.
* Two machines are both *offline* and have the *same* name. You turn both of them on. One of them should notice the conflict and correct it.
* You give *three* machines the same name. You should see something like `MyRobot`, `MyRobot #2`, and `MyRobot #3`.
* You have one OT-2 connected to your computer over both a wired and wireless connection. It should not conflict with itself or change its name on its own.

## After testing

When you want to go back to `edge` after testing this PR, **do not do `git switch edge && make -C update-server push`.** It will soft-brick your robot because of https://github.com/Opentrons/opentrons/issues/10582. Instead, download an `edge` build of `ot2-system.zip` and push it through the Opentrons App.

# Risk assessment

High.

* Screwing stuff up in this part of the codebase can soft-brick the robot by crashing `update-server`.
* Screwing stuff up in this part of the codebase can soft-brick the robot by by accidentally breaking some nuance of robot discovery.
* This part of the codebase has spotty type-checking and unit testing coverage.
* We're handling interactions with system processes that we don't own and that we can only test manually.
